### PR TITLE
feat: zip build exclusions support all globs

### PR DESCRIPTION
**Pull request by patrzhan**
**Issue #, if available:**
https://github.com/aws-greengrass/aws-greengrass-gdk-cli/issues/143
**Description of changes:**
Update zip build to accept full globs to match with when picking files/directories to exclude from the build. In GDK 1.4.0 and earlier, glob-like patterns were provided to match individual file names against, regardless of the subdirectory level. In PR: https://github.com/aws-greengrass/aws-greengrass-gdk-cli/pull/253, this was changed to only exclude items at the root level. This change supports full glob notation including directories, so that items that should only be excluded from root (i.e. `test*`) can do so, but items that should still be excluded from all subdirectories such as node_modules folders can also be excluded.

This also means the defaults have changed. The default globs for exclusion have been set to `test*`, `**/.*`, and `**/node_modules`, meaning that test* is matched only against the root directory, but hidden files and node_modules are excluded at any level by default. This also means that any custom patterns provided in the gdk-config.json file may need to be changed, as the behavior may have changed.

**Why is this change necessary:**
Bugfix for the linked issue while maintaining support for exclusion in any subdirectory level.

**How was this change tested:**
Manually tested changes with a sample project to confirm that expected directories and files were excluded based on various provided globs. Also updated unit/integration tests to test the functionality that is changed.

**Any additional information or context required to review the change:**
Since this feature impacts default values and may change behavior of existing projects that upgrade to this version, input is needed on how to mitigate impact if necessary.

**Checklist:**
- [ ] Updated the README if applicable
- [ ] Updated or added new unit tests
- [ ] Updated or added new integration tests
- [ ] Updated or added new end-to-end tests
- [ ] If your code makes a remote network call, it was tested with a proxy
- [ ] You confirm that the change is backwards compatible

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: gdk/common/utils.py [hunk 1]
```diff
@@ -152,6 +154,15 @@ def is_recipe_size_valid(file_path):
     return file_size <= MAX_RECIPE_FILE_SIZE_BYTES, file_size
 
 
+def generate_ignore_list_from_globs(root_directory, globs):
```
Comment (saranyailla) [inline gdk/common/utils.py:157]: Why is it not inside Zip? Who else is going to need this?

Comment (patrzhan) [reply gdk/common/utils.py]: Moved it to Zip file

File: gdk/build_system/Zip.py [hunk 2]
```diff
@@ -41,19 +43,11 @@ def build(self, **kwargs):
...
-                if str(path) == str(root_directory_path):
-                    return {
-                        name
-                        for pattern in self.get_ignored_file_patterns(project_config)
-                        for name in fnmatch.filter(names, pattern)
-                    }
-                return set()
-
-            shutil.copytree(
+            custom_copytree.copytree(
```
Comment (MikeDombo) [inline gdk/build_system/Zip.py:46]: why not keep using shutil, but with a custom filter that looks at the full path, rather than only the filename?

Comment (patrzhan) [reply gdk/build_system/Zip.py]: Yea at first I thought this wouldn't work but realized it would work after all. See latest commit.

**Review by MikeDombo — APPROVED** [review]

**Review by saranyailla — APPROVED** [review]

**Pull request merged**